### ISSUE TITLE
`no-octal-literal` underdelivers

### ISSUE THREAD

**Issue by IllusionMH** [microsoft/tslint-microsoft-contrib#653]
### Bug Report

-   **`tslint-microsoft-contrib` version**: 6.0.0-beta (`master` branch)
-   **TSLint version**: 5.1.1
-   **TypeScript version**: 3.1.1
-   **Running TSLint via**: CLI

During migation of tests I've noticed some strange behviors in this rule.

#### TypeScript code being linted

```tsx
// first case
var x1 = 'Sample text \0111'; // longer than octal

// second case
var x2 = 'Sample text \0';

// third case
var x3 = 'Sample text \777';

// fourth case
var x4 = 0o123;
```
First case is part of [passing test](https://github.com/Microsoft/tslint-microsoft-contrib/blob/4b911106ba4efa13996f05ada4a2ec7450330bc1/test-data/NoOctalLiteral/NoOctalLiteralTestInput-passing.ts#L3)

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-octal-literals": true
    }
}
```

#### Actual behavior

First case: No error
Second case: `Octal literals should not be used: \0`
Third case: `Octal literals should not be used: \777`
Fourth case: No error

#### Expected behavior

**First case:**
Error because this string actually contains octal escape `\011`, and `1` is just following character.
And exactly this string will throw in `strict mode`
```js
(function() {
    'use strict';
    var x1 = 'Sample text \0111'; // longer than octal
})()
```

**Second case:**
Should be no error because `\0` is allowed in `strict mode` - it is NULL character (unless followed by decimal character)

<details>
<summary>Interesting observation</summary>
<div>

`'\09'` will throw because of octal escape in Edge & Opera (even if `9` can't be part of octal escape, but still considered octal, details [here](https://mathiasbynens.be/notes/javascript-escapes))

`` `\09` `` will throw in Opera, but not in Edge

</div>

</details>

**Third case:**
Should be: `Octal literals should not be used: \77`
Because octal escapes should not be bigger than `0o377` (`255`) therefore only `\77` is escape and `7` is next character. `'\777' === '?7'`

**Fourth case:**
There are just no checks for octal literals (a.k.a. octal numeric literals)  at all, however rule description is: "Do not use **octal literals** or escaped octal sequences." and code clearly only  implements second part partially (it still allows escapes that will throw in `strict mode`).

I think that checks for escapes should be strict by default with proper handling of `\0`, but for numbers I see two options:

1. Add check for octal number literals by default.

1. Add `check-numbers` option that will enable check on octal literals.

What do you think?

**Comment by IllusionMH** [comment]
Another problem spotted is errors like `Octal literals should not be used: \-035`.
Ignoring "literal part" - octal escapes cannot include `-` character.
In test string `return "Sample text \-035";` it would be just escaped `-` character.
```
> (function() {'use strict'; return 'Sample text \-035';})()
< "Sample text -035"

> (function() {'use strict'; return 'Sample text \035';})()
< Uncaught SyntaxError: Octal escape sequences are not allowed in strict mode.
```

I think this should be considered during fixing problems mentioned in original post.

**Comment by IllusionMH** [comment]
After second and third thoughts I think there is alternative approach:
1. Implement new separate rule `no-octal-escapes` (as I can see there is no such rule in [tslint-eslint-rules](https://www.npmjs.com/package/tslint-eslint-rules) but ESLint has one).
1. Mark `no-octal-literal` as deprecated and remove eventually (not sure if in 7.0 or 8.0).

Reasoning not to check number literals as I suggested earlier:
1. Problematic cases were `644 !== 0644` and `90 === 090`.
However TS already warns about problematic cases (when digits after 0 are 1-7).
```
error TS1085: Octal literals are not available when targeting ECMAScript 5 and higher. Use the syntax '0o644'.
```
Not sure if linter rule should go against this recommendation and flag also `0o664`

2. `0o644` is explicit and doesn't produce problems. Only readability is a bit hurt in case of `0O644` but in editor with adequate font this shouldn't be a big problem.

Reasoning to implement `no-octal-escapes`:
1. It is runtime error in strict mode that better to catch earlier.
1. TS has long standing issue with marking octal escapes as errors https://github.com/Microsoft/TypeScript/issues/396
1. Similarity with ESLint

And following question where `no-octal-escapes` should be implemented. I think that core package is a best place, but can be any of `tslint`, `tslint-microsoft-contrib` or `tslint-eslint-rules`.

**Comment by IllusionMH** [comment]
Yes, create new rule in code and deprecate one in this repo.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #876. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!
